yt/visualization/volume_rendering/off_axis_projection.py

```diff
@@ -104,20 +106,66 @@ def off_axis_projection(data_source, center, normal_vector,
     >>> write_image(np.log10(image), "offaxis.png")
 
     """
-
     if method not in ['integrate','sum']:
         raise NotImplementedError("Only 'integrate' or 'sum' methods are valid for off-axis-projections")
 
     if interpolated is True:
         raise NotImplementedError("Only interpolated=False methods are currently implemented for off-axis-projections")
+    
+    data_source = data_source_or_all(data_source)
 
+    item = data_source._determine_fields([item])[0]
 
-    data_source = data_source_or_all(data_source)
     # Sanitize units
     if not hasattr(center, "units"):
         center = data_source.ds.arr(center, 'code_length')
     if not hasattr(width, "units"):
         width = data_source.ds.arr(width, 'code_length')
+
+    if hasattr(data_source.ds, '_sph_ptype'):
+        ptype = data_source.ds._sph_ptype
+
+        fi = data_source.ds.field_info[item]
+
+        raise_error = False
+
+        if fi.alias_field:
+            if fi.alias_name[0] != ptype:
+                raise_error = True
+        else:
+            if fi.name[0] != ptype:
+                raise_error = True
+
+        if raise_error:
+            raise RuntimeError(
+                    "Can only perform off-axis projections for SPH fields, "
+                    "Received '%s'" % (item,)
+                    )
+
+        buf = np.zeros((resolution[0], resolution[1]), dtype='float64')
+
+        x_min = center[0] - width[0] / 2
+        x_max = center[0] + width[0] / 2
+        y_min = center[1] - width[1] / 2
+        y_max = center[1] + width[1] / 2
+        z_min = center[2] - width[2] / 2
+        z_max = center[2] + width[2] / 2
+
+        bounds = [x_min, x_max, y_min, y_max, z_min, z_max]
+        for chunk in data_source.chunks([], 'io'):
+            off_axis_projection_SPH(chunk[ptype, "particle_position_x"].to('code_length').d,
+                                    chunk[ptype, "particle_position_y"].to('code_length').d,
+                                    chunk[ptype, "particle_position_z"].to('code_length').d,
+                                    chunk[ptype, "particle_mass"].to('code_mass').d,
+                                    chunk[ptype, "density"].to('code_density').d,
+                                    chunk[ptype, "smoothing_length"].to('code_length').d,
+                                    bounds,
+                                    center.to('code_length').d,
+                                    width.to('code_length').d,
+                                    chunk[item],
+                                    buf,
+                                    normal_vector)
+        return buf
```

Comment:
You should be returning an `ImageArray` with the correct units, see around line 248 below.

Reply:
Hmm, so unless you used a weight field projected density should have units of `g/cm**2`. See right here: `https://github.com/bqia0/yt/blob/3ddeaa27bc0f5d7dd01aefb63d91f92fa1986b51/yt/visualization/volume_rendering/off_axis_projection.py#L277`, although I'm not sure if we need to do something like that, or if we just need to get the projected units. I'd need to look closer at how you implemented the SPH off-axis projection wrapper to tell you exactly how that should be handled.

The info dict looks great, my only suggestion would be to change the type to "SPH smoothed projection"